pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+}
+
+// deleteFinalizer remove the finalizer from the instance CRD.
+func (r *InstanceReconciler) deleteFinalizer(ctx context.Context) error {
+	// remove our finalizer from the list and update it.
+	controllerutil.RemoveFinalizer(r.instance, FinalizerName)
+	return r.Update(ctx, r.instance)
+}
+
+// createOrUpdateResource uses for creates or updates k8s resource, in our case used only for deployments.
+func (r *InstanceReconciler) createOrUpdateResource(ctx context.Context, object client.Object) error {
+	// Use CreateOrUpdate to create or update the PVC for cl-controlplane
+	_, err := controllerutil.CreateOrUpdate(ctx, r.Client, object, func() error {
+		createTime := object.GetCreationTimestamp()
+		if createTime.IsZero() {
+			r.Logger.Infof("Create resource %v Name: %s Namespace: %s", reflect.TypeOf(object), object.GetName(), object.GetNamespace())
+		}
+		return nil
+	})
+
+	if err != nil {
+		r.Logger.Error(err)
+	}
+	return err
+}
+
+// createResource uses for creates k8s resource.
+func (r *InstanceReconciler) createResource(ctx context.Context, object client.Object) error {
+	err := r.Get(ctx, types.NamespacedName{Name: object.GetName(), Namespace: object.GetNamespace()}, object)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+
+		r.Logger.Infof("Create resource %s Name: %s Namespace: %s", reflect.TypeOf(object), object.GetName(), object.GetNamespace())
+		return r.Create(ctx, object)
+	}
+
+	return nil
+}
+
+// deleteAllResources delete all the ClusterLink resource.
+func (r *InstanceReconciler) deleteAllResources(ctx context.Context) error {
+	// Delete deployments
+	if err := r.deleteResource(ctx, DataPlaneName, &appsv1.Deployment{}); err != nil {
+		return err
+	}
+
+	if err := r.deleteResource(ctx, ControlPlaneName, &appsv1.Deployment{}); err != nil {
+		return err
+	}
+	// Delete services
+	if err := r.deleteResource(ctx, DataPlaneName, &corev1.Service{}); err != nil {
+		return err
+	}
+
+	if err := r.deleteResource(ctx, ControlPlaneName, &corev1.Service{}); err != nil {
+		return err
+	}
+
+	// Delete PVC
+	if err := r.deleteResource(ctx, ControlPlaneName, &corev1.PersistentVolumeClaim{}); err != nil {
+		return err
+	}
+
+	// Delete rules
+	if err := r.deleteResource(ctx, ControlPlaneName, &rbacv1.ClusterRole{}); err != nil {
+		return err
+	}
+
+	if err := r.deleteResource(ctx, ControlPlaneName, &rbacv1.ClusterRoleBinding{}); err != nil {
+		return err
+	}
+
+	// Delete ingress
+	if err := r.deleteResource(ctx, IngressName, &corev1.Service{}); err != nil {
+		return err
+	}
+
+	if r.clNamespace == InstanceNamespace {
+		if err := r.deleteResource(ctx, ControlPlaneName, &corev1.Namespace{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// deleteResource delete a k8s resource.
+func (r *InstanceReconciler) deleteResource(ctx context.Context, name string, object client.Object) error {
+	err := r.Get(ctx, types.NamespacedName{Name: name, Namespace: r.clNamespace}, object)
```

Comment:
If the intent is to delete always, why do you need a Get first?

Reply:
To save API calls, and prevent an error in the delete command.